### PR TITLE
This PR introduces a runtime available type code suitable for use in buffer_t or other contexts.

### PR DESCRIPTION
**Pull request by zvookin**
I have a need for this in some code I am using as well as it supporting the requirement for full type info in the next version of buffer_t. This may not be exactly right, but I wanted to the discussion started. Particular questions are how much C++ support to include in HalideRuntime.h and whether to duplicate runtime types with compile time types in the way halide_type_t and Halide::Type do. In this case I think it is productive to do so. (We have to keep some amount of support in the Halide namespace to keep end user code compiling as Int(8) and company are used in Halide declarations fairly frequently. I believe Halide::Type is used enough that removing it would be a fairly large breaking change.)

I chose to use bitfields here because using uint8_t as a field type makes the field into an unsigned char, which C++ treats as a character type when writing to streams, etc. That is unexpected and I ran into it in a couple places doing the conversion so I went with this approach. I could be talked out of doing it this way.

As the bits and width fields are no longer signed, the wild card value for these had to be changed from -1 to 0. There is some slight possibility this will result in expr_match reporting matches that would not have happened before. I did look for this and thing the compiler is safe.

The sizing of bits and width allows for up to 128-bit scalars and up to 32k vector width. I chose this partitioning as 1-bit (bool) vectors of >256 elements already exist. I would like to keep the code fitting in 32-bits to keep the amount of space it takes up in buffer_t small.

### CODE REVIEW

File: src/runtime/HalideRuntime.h [hunk 1]
```diff
@@ -490,6 +494,37 @@ typedef enum halide_type_code_t {
...
+ * unsigned ints, or floats of various bit-widths (the 'bits'
+ * field). Can also be vectors of the same (by setting the 'width'
+ * field to something larger than one). This struct is should be
+ * exactly 32-bits in size. */
+struct halide_type_t {
+    /** The basic type code: signed integer, unsigned integer, or floating point. */
+#if __cplusplus >= 201103L
+    HALIDE_ATTRIBUTE_ALIGN(1) halide_type_code_t code; // halide_type_code_t
+#else
+  HALIDE_ATTRIBUTE_ALIGN(1) unsigned int code : 8; // halide_type_code_t
```
Comment (steven-johnson) [inline src/runtime/HalideRuntime.h]: Wonky indentation

**Comment by steven-johnson** [comment]
0 as the wildcard value seems fine and sensible.

Adding C++ support to HalideRuntime.h seems fine with the caveat that it should be syntactic sugar only -- i.e., I'm certain we want to maintain an extern-C-only API for compatibility with other language bindings.

Bitfields make me nervous; this has been a notorious point of compiler incompatibility in the past. (Does the language runtime make any guarantees about memory layout?) I'm not convinced that the confusion-with-char-for-streams is enough to justify this (can you expand a bit more to convince me?)

**Comment by kgnk** [comment]
When I have to stream `uint8_t` values, I `static_cast<int>()` them.

**Comment by abadams** [comment]
Based on the travis failure, it looks like there are places in the code that assume the width of a type is an int. So it needs an int cast at least. Maybe it's worth biting the bullet and hiding the width, type code, and bits behind methods that return int (and then not inheriting Type from halide_type_t). There are places where we make a copy of a type and then mess with its fields (e.g. bits*=2) to make a new type, and those would have to be replaced with things like Type t(other.code(), other.bits()*2, other.width()); 

Maybe that's good. immutable objects are nice. We could add getters but no setters. If that's beyond the scope of what you want to do here, then I'm OK with the current approach too once fixed.

File: src/runtime/HalideRuntime.h [hunk 2]
```diff
@@ -490,6 +494,37 @@ typedef enum halide_type_code_t {
...
+#endif
+
+    /** The number of bits of precision of a single scalar value of this type. */
+  HALIDE_ATTRIBUTE_ALIGN(1) unsigned int bits : 8;
+
+    /** How many elements (if a vector type). Should be 1 for scalar types. */
+  HALIDE_ATTRIBUTE_ALIGN(2) unsigned int width : 16;
+
+#ifdef __cplusplus
+    halide_type_t(halide_type_code_t code, uint8_t bits, uint16_t width = 1) : code(code), bits(bits), width(width) { }
```
Comment (abadams) [inline src/runtime/HalideRuntime.h]: Why explicit bit sizes instead of uint8 and uint16?

Also, I doubt we need this c++ block. We can just use brace initialization for the first constructor, and aren't the other two functions the same as the default copy and assignment operators? Or am I missing something...

File: src/runtime/HalideRuntime.h [hunk 3]
```diff
@@ -765,4 +800,82 @@ extern double halide_float16_bits_to_double(uint16_t);
 } // End extern "C"
 #endif
 
+#ifdef __cplusplus
```
Comment (abadams) [inline src/runtime/HalideRuntime.h]: This C++ block looks like useful syntactic sugar to me. I think it's worth having in HalideRuntime.h

Comment (zvookin) [inline src/runtime/HalideRuntime.h]: I'm all for encapsulating at the Type level. I was likely erring on the
side of not changing the compiler code too much because it seemed
indicative of a bad change, but given this input I'll pursue that approach.

I think the constructors were necessary to have Type inherit from
halide_type_t.
If we do the encapsulation, those can be nuked. Bit fields are in the same
boat as I did that because a test failed do to a change in behavior with no
compile error and that bothered me. (C++'s type system conflating
characterness with bit size and then having overloading is really not a
good thing.)

**Comment by zvookin** [comment]
I've pushed a version of this with better encapsulation for the Halide::Type type. Please take a look. In particular, are "with_code", "with_bits" and "with_width" the right approach and naming?

I went with unsigned types for the bits() and width() methods. This does not seem to have required more casts, but perhaps they will turn up on another platform.

**Comment by zvookin** [comment]
I also made raise_to_integer_power take an int64_t as otherwise it could possibly truncate silently and it seems this is the easiest way to fix that.

**Comment by kgnk** [comment]
It wasn't so silent. I had to use `-isystem halide/include` to silence that `raise_to_integer_power` truncation warning in our projects.

File: src/Type.h [hunk 4]
```diff
@@ -20,81 +20,92 @@ struct Expr;
...
+    static constexpr halide_type_code_t Int = halide_type_int;
+    static constexpr halide_type_code_t UInt = halide_type_uint;
+    static constexpr halide_type_code_t Float = halide_type_float;
+    static constexpr halide_type_code_t Handle = halide_type_handle;
+    // @}
 
-    /** How many elements (if a vector type). Should be 1 for scalar types. */
-    int width;
+    /** The number of bytes required to store a single scalar value of this type. Ignores vector width. */
+    int bytes() const {return (bits() + 7) / 8;}
```
Comment (steven-johnson) [inline src/Type.h]: should this return an unsigned (or size_t)?

File: src/Type.h [hunk 5]
```diff
@@ -20,81 +20,92 @@ struct Expr;
...
+        return Type(code(), new_bits, width());
+    }
 
-    Type& operator=(const Type& that) {
-        this->code = that.code;
-        this->bits = that.bits;
-        this->width = that.width;
-        return *this;
+    /** Return Type with same type code and number of bits, but new_width for the vector width. */
+    // TODO(zalman): Should this be called broadcast?
```
Comment (steven-johnson) [inline src/Type.h]: with_width() sounds reasonable to me.

File: src/runtime/HalideRuntime.h [hunk 6]
```diff
@@ -467,13 +467,16 @@ extern int halide_error_debug_to_file_failed(void *user_context, const char *fun
...
-
 /** Types in the halide type system. They can be ints, unsigned ints,
  * or floats (of various bit-widths), or a handle (which is always pointer-sized).
  * Note that the int/uint/float values do not imply a specific bit width
  * (the bit width is expected to be encoded in a separate value).
  */
-typedef enum halide_type_code_t {
+typedef enum halide_type_code_t 
+#if __cplusplus >= 201103L
+: uint8_t
```
Comment (steven-johnson) [inline src/runtime/HalideRuntime.h]: If we're going to specialize for C++11, should we (also) make it "enum class"?

File: src/runtime/HalideRuntime.h [hunk 7]
```diff
@@ -490,6 +493,33 @@ typedef enum halide_type_code_t {
...
+  HALIDE_ATTRIBUTE_ALIGN(2) uint16_t width;
+
+#ifdef __cplusplus
+    halide_type_t(halide_type_code_t code, uint8_t bits, uint16_t width = 1) : code(code), bits(bits), width(width) { }
+
+    /** Size in bytes for a single element, even if width is not 1, of this type. */
+  size_t bytes() { return (bits + 7) / 8; };
+#endif
+};
+
```
Comment (steven-johnson) [inline src/runtime/HalideRuntime.h]: Should we also add

   static_assert(sizeof(halide_type_t) == 4);

?

File: src/runtime/cache.cpp [hunk 8]
```diff
@@ -139,7 +144,7 @@ WEAK bool CacheEntry::init(const uint8_t *cache_key, size_t cache_key_size,
     for (size_t i = 0; i < key_size; i++) {
         key[i] = cache_key[i];
     }
-    for (int32_t i = 0; i < tuple_count; i++) {
+    for (uint32_t i = 0; i < tuple_count; i++) {
```
Comment (steven-johnson) [inline src/runtime/cache.cpp]: Using "unsigned" (vs uint32_t) for loop counters may be slightly more performant on 64-bit architectures.

File: src/Type.h (hunk 5 above)
Comment (abadams) [inline src/Type.h]: The name is fine, but it's the same thing as vector_of. Maybe delete vector_of. There's also element_of, but that seems worth keeping.

File: src/Type.h [hunk 9]
```diff
@@ -20,81 +20,92 @@ struct Expr;
...
 
-    Type(TypeCode _code, int _bits, int _width) : code(_code), bits(_bits), width(_width) {}
+    Type(const Type &that) : type(that) {}
+
+    Type(const halide_type_t &that) : type(that) {}
+
+    operator halide_type_t() const { return type; }
+
+    halide_type_code_t code() const { return type.code; }
+    int bits() const { return type.bits; }
```
Comment (abadams) [inline src/Type.h]: Don't forget doxygen comments for these new methods

File: src/IROperator.cpp [hunk 10]
```diff
@@ -567,7 +567,7 @@ Expr halide_erf(Expr x_full) {
     return result;
 }
 
-Expr raise_to_integer_power(Expr e, int p) {
+Expr raise_to_integer_power(Expr e, int64_t p) {
```
Comment (steven-johnson) [inline src/IROperator.cpp]: out of curiosity, why is this necessary? (ie where are we calling this with int64)

File: src/runtime/HalideRuntime.h [hunk 11]
```diff
@@ -490,6 +493,33 @@ typedef enum halide_type_code_t {
...
 
+/** A runtime tag for a type in the halide type system. Can be ints,
+ * unsigned ints, or floats of various bit-widths (the 'bits'
+ * field). Can also be vectors of the same (by setting the 'width'
+ * field to something larger than one). This struct is should be
+ * exactly 32-bits in size. */
+struct halide_type_t {
+    /** The basic type code: signed integer, unsigned integer, or floating point. */
+#if __cplusplus >= 201103L
+  HALIDE_ATTRIBUTE_ALIGN(1) halide_type_code_t code; // halide_type_code_t
```
Comment (abadams) [inline src/runtime/HalideRuntime.h]: wonky indentation in this struct

**Comment by steven-johnson** [comment]
LGTM

File: src/runtime/HalideRuntime.h (hunk 6 above)
Comment (abadams) [inline src/runtime/HalideRuntime.h]: Wouldn't that require halide_type_code_t::halide_type_int? That seems pretty verbose.

Comment (steven-johnson) [inline src/runtime/HalideRuntime.h]: Hmm, yeah. Not the right idiom for a C-like enum anyway. Nevermind.

File: src/Type.h (hunk 4 above)
Comment (zvookin) [inline src/Type.h]: Likely "yes," but changing existing functionality in Type.h to use unsigned types is going to require either ignoring type sign mistmatches or fixing a lot of code. I opened https://github.com/halide/Halide/issues/979 to make an umbrella task to solve this in the broad sense and I made the version of bytes on halide_type_t return size_t so new uses get the right thing. I'd rather not tackle fixing the entire issue as part of this PR however.

File: src/Type.h (hunk 5 above)
Comment (dsharletg) [inline src/Type.h]: I think vector_of is more descriptive, but it's also not great because the name suggests you'll get a vector of vectors if called on a type that is already a vector. If we kept that name, we should add an assert that the current vector width is 1.

While we're at this change, I've been briefly confused in the past when I attempted to use "width" to mean "integer width" and not "vector width". Maybe width should be vector_width? i.e. with_vector_width.

That was a bit of a ramble. To clarify, I think we should either keep vector_of, but add an assert that the type is scalar, or we should keep with_width, but consider renaming it to with_vector_width (and same for width()).

Comment (zvookin) [inline src/Type.h]: Sigh, I didn't even see vector_of. I think there are places where I changed "width = ..." to a with_width form that probably should have been using vector_of in the first place. I like that name as I feel "width" in general requires internalizing that inside Halide "width" is the number of items in a vector, except that it isn't when you're writing Halide code itself (where it is a synonym for "extent(0)").

The only issue with "vector_of" is making it clear that it replaces the width when called on a Type with width != 1. Not sure if there is a name that is still terse that makes that clear. I guess with_width makes that clear...

File: src/IROperator.cpp (hunk 10 above)
Comment (zvookin) [inline src/IROperator.cpp]: Immediates can now be 64-bit. See: https://github.com/halide/Halide/blob/master/src/IROperator.h#L1191

File: src/Type.h (hunk 5 above)
Comment (zvookin) [inline src/Type.h]: How do folks feel about changing "width" to "lanes" through out the code?

Comment (abadams) [inline src/Type.h]: +1

It's an unambiguous widely used word, and there's some precedent in the code in the functions in Deinterleave.h

**Comment by abadams** [comment]
Seems like a good candidate PR for typos breaking the build in corner cases, so I kicked off the full test on a few platforms: 
http://buildbot.halide-lang.org:8010/builders/linux-64-gcc48-trunk/builds/663
http://buildbot.halide-lang.org:8010/builders/mac-64-trunk/builds/675
http://buildbot.halide-lang.org:8010/builders/win-64-trunk/builds/615

**Comment by abadams** [comment]
visual studio 2013 appears to be unhappy with the constexpr stuff :(
